cpp/src/arrow/engine/substrait/options.h

```diff
@@ -0,0 +1,55 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// This API is EXPERIMENTAL.
+
+#pragma once
+
+namespace arrow {
+namespace engine {
+
+/// How strictly to adhere to the input structure when converting between Substrait and
+/// Acero representations of a plan. This allows the user to trade conversion accuracy
+/// for performance and lenience.
+enum class ConversionStrictness {
+  /// Prevent information loss by rejecting incoming plans that use features or contain
+  /// metadata that cannot be exactly represented in the output format in a way that
+  /// will round-trip. Relations/nodes must map one-to-one.
+  PEDANTIC,
+
+  /// When an incoming plan uses a feature that cannot be exactly represented in the
+  /// output format, attempt to emulate that feature as opposed to immediately
+  /// rejecting the plan. For example, a Substrait SortRel with a complex sort key
+  /// expression may be emulated using a project-order-project triple. Relations/nodes
+  /// will thus map one-to-many.
```

Comment:
I'm not sure what you mean by "supplementary node." Can you give an example?

If you mean for example a no-op project node, my intention for PRESERVE_STRUCTURE is to keep that no-op project node in there. Removing it would be an optimization-esque thing and thus belongs in BEST_EFFORT.

Reply:
Looking through the current nodes I can't come up with a good concrete example.  Perhaps something like a sort node.  We don't support mid-plan sort in Acero today (because we can't maintain the ordering after the node) but we do support sorting sinks (because there is no "after the node").  So it's possible we could have a RunSubstraitSorted function that allows plans that end in a sort node even though the sort node itself wouldn't be in the actual plan that gets run (just the sink would be sorted).

It's a bit of a stretch though.  Either way, I think this is more of an implementation detail and not something to concern users.